tooltips: Replace URL title tags with Tippy tooltips.

reharsh opened the pull request:
This PR:
- replaces title tags with Tippy tooltips for named URLs with no delays, allowing users to see the actual URL immediately and providing protection against fraudulent URLs.
- Tippy tooltips for plain URLs, internal Zulip links (channels / topics / messages), user uploads with a **long hover delay**.

fixes: #32973

**Screenshots and screen captures:**
wrong links:
<img width="662" alt="Screenshot 2025-01-14 at 3 49 04 PM" src="https://github.com/user-attachments/assets/c1c70556-bb41-44f4-a36a-1ac1cbf5b02a" />
update: <img width="342" alt="Screenshot 2025-02-03 at 10 10 07 PM" src="https://github.com/user-attachments/assets/82ccd0b7-8e0f-4131-88e2-424208297b4b" />



update with internal links in canonical form:

https://github.com/user-attachments/assets/23523ec0-892a-4abc-91fe-1f55985989bb





<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

reharsh commented:
@alya, could you please check this out as well? thanks!

alya commented:
The issue said to do this for named links, but I see it for plain URLs as well.

alya commented:
I think what we actually want is Tippy tooltips for plain URLs too, but with a long hover delay. And I would do the same for internal Zulip links (channels / topics / messages). It seems weird to leave those as HTML titles.

reharsh commented:
@alya, added tippy tooltips for other elements as well (with long hover delay). It does look much better now—thanks!
updated screen recording, please check.

abhishek07ma commented:
@zulipbot claim

alya commented:
Thanks! Looking at the screen capture, we're going to need to refine it further for internal links:

- For internal links that are already in the standard format (#channel, #channel > topic, or message links like #sandbox > blue nas de-duping @ 💬), and point to the expected place, no tooltip at all (no additional info to communicate).
- For internal links that aren't in the standard format (e.g., "said" links in quote-and-reply, or whatever other label a user chooses, or #channel1 that actually points to #channel2, etc.), let's try explaining where the link goes, not the URL:
  > "Go to {canonical format for location}
We should try having no delay on these tooltips.

reharsh commented:
@alya, just a gentle reminder for this one.

alya commented:
You need to post a comment regarding the status of the PR for reviewers to know that they need to take a look.

Has my feedback been addressed? Have screenshots been updated accordingly?

reharsh commented:
sure, @alya -- will keep it in mind in the future

updates:
- [x] only external named links and links that don't point to where they're expected show url tooltip with no delay
- [x] Internal links that are already in standard format and point to expected place don't show tooltip at all.
- [x] Internal links that are not in standard format or don't point to expected place -> show tooltip in **canonical form** like: (Go to # design > tippy tooltip) (with delay)

above feedback has been addressed and screenshots are updated accordingly. thanks

alya commented:
Thanks! @sayamsamal are you up for reviewing this one? I haven't re-tested.

sayamsamal commented:
Hello @reharsh, could you rebase this PR so I can review this?

reharsh commented:
Hii @sayamsamal, just a gentle reminder for this pr.

sayamsamal commented:
Thanks @reharsh, sorry for the delay. Reviewing the PR now.

sayamsamal commented on web/src/postprocess_content.ts:
```diff
@@ -74,35 +75,22 @@ export function postprocess_content(html: string): string {
+        // canonical form via tooltip with some delay.
+        if (
+            url.origin !== window.location.origin &&
+            elt.textContent?.trim() !== elt.href.replace(/\/$/, "")
```
Do we really need this check over here? I guess it would be more consistent if we just showed the tooltips regardless of whether the text and link match.

reharsh replied:
I have added an explanation in the comment there, is it okay?

sayamsamal commented on web/src/postprocess_content.ts:
```diff
@@ -74,35 +75,22 @@ export function postprocess_content(html: string): string {
+            url.origin !== window.location.origin &&
+            elt.textContent?.trim() !== elt.href.replace(/\/$/, "")
+        ) {
+            elt.setAttribute("class", "inline_url");
```
If we are confirmed by this point that the URL is infact an external url, then let's also let the class name include the word `external`. Maybe `external_inline_url`.

reharsh replied:
done!

sayamsamal commented on web/src/postprocess_content.ts:
```diff
@@ -74,35 +75,22 @@ export function postprocess_content(html: string): string {
+            url.origin === window.location.origin &&
+            !STANDARD_INTERNAL_URLS.has(elt.getAttribute("class") ?? "")
+        ) {
+            elt.setAttribute("class", "internal_named_link");
```
Using class name to store this information doesn't sound intuitive since there is always the risk of modifying the classes on an element by chance. How about using `data-message-link-type` attribute instead?

reharsh replied:
done!

sayamsamal replied:
I'm still seeing the usage of class for external_inline_url. If we are using a `data-*` attribute, we might as well use it for all the cases.

sayamsamal commented:
I see `message-link` in the `STANDARD_INTERNAL_URLS` set but I don't see the support for message links. Maybe it was a result of the recent changes in the message links? If so, let's update the PR to support them.

sayamsamal commented:
@reharsh I have left you some feedback on the code, please tag me again when the PR is ready for another round of review.

reharsh commented on web/src/postprocess_content.ts:
```diff
@@ -74,35 +75,22 @@ export function postprocess_content(html: string): string {
+            elt.setAttribute("class", "inline_url");
+        } else if (
+            url.origin === window.location.origin &&
+            !STANDARD_INTERNAL_URLS.has(elt.getAttribute("class") ?? "")
```
Actually for standard format links we don't show tooltips as alya suggested [here](https://github.com/zulip/zulip/pull/33039#issuecomment-2605598608)

sayamsamal commented on web/src/message_list_tooltips.ts:
```diff
@@ -349,6 +351,64 @@ export function initialize(): void {
+                // We show the tooltip in canonical form for the
+                // destination internal link is pointing to.
+                const syntax = (): string => {
+                    const syntaxt_text = try_stream_topic_syntax_text(
+                        window.location.origin + url!,
+                    );
+                    const inputString = syntaxt_text;
+
+                    const regex = /#\*\*([^>]+?)(?:>([^*]*))?\*\*/;
+                    const match = inputString?.match(regex);
+
```
We have `hash_util.decode_stream_topic_from_url()` which does the extraction of topic and channel for us, so I say we use that instead of using this regex which might not be the most robust solution here.

reharsh replied:
done!

sayamsamal commented:
The tooltip on the uploaded file sounds wrong, we should probably be sticking the previously used: `Download <filename.extension>`
![Screenshot 2025-04-15 at 2 25 01 AM](https://github.com/user-attachments/assets/abd46851-d17c-458e-9021-f3b1a9eab30b)

sayamsamal commented on web/src/postprocess_content.ts:
```diff
@@ -149,35 +150,25 @@ export function postprocess_content(html: string): string {
+            url.origin === window.location.origin &&
+            !STANDARD_INTERNAL_URLS.has(elt.getAttribute("class") ?? "")
+        ) {
+            elt.dataset.messageLinkType = "internal_named_link";
```
Any reason to use `dataset` over here than using the `setAttribute()` function we have been using at other places? `setAttribute()` is likely more performant. https://jsperf.app/html5-data-attribute-dataset-vs-getattribute/4.

sayamsamal commented:
@reharsh I have left you some more feedback.

reharsh commented:
<img width="504" alt="image" src="https://github.com/user-attachments/assets/3b8234a2-0811-4f31-b4ba-59a6682051b7" />

it does show `download {file}`.

sayamsamal commented on web/src/message_list_tooltips.ts:
```diff
@@ -2,18 +2,21 @@ import $ from "jquery";
 import * as rows from "./rows.ts";
 import {message_edit_history_visibility_policy_values} from "./settings_config.ts";
 import {realm} from "./state_data.ts";
+import {get_stream_name_from_id} from "./stream_data.ts";
```
Please use [namespace imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#namespace_import), instead of [named imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#named_import) — it helps with the code readability.

sayamsamal commented on web/src/message_list_tooltips.ts:
```diff
@@ -2,18 +2,21 @@ import $ from "jquery";
 import render_narrow_tooltip from "../templates/narrow_tooltip.hbs";
 
 import * as compose_validate from "./compose_validate.ts";
+import {decode_stream_topic_from_url} from "./hash_util.ts";
```
Please use [namespace imports](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/import#namespace_import) — https://github.com/zulip/zulip/pull/33039/files#r2060258348

sayamsamal commented on web/src/message_list_tooltips.ts:
```diff
@@ -349,6 +352,60 @@ export function initialize(): void {
+                    if (stream_topic_name !== null) {
+                        const channel = get_stream_name_from_id(stream_topic_name.stream_id);
+                        const topic = stream_topic_name.topic_name;
+                        return topic ? `# ${channel} > ${topic}` : `# ${channel}`;
```
Let's remove the space between `#` and the channel name.

reharsh replied:
<img width="363" alt="image" src="https://github.com/user-attachments/assets/9794883c-38e9-4c3f-8e45-a2119dff2b1f" />

done!

sayamsamal commented on web/src/message_list_tooltips.ts:
```diff
@@ -349,6 +352,60 @@ export function initialize(): void {
+                // happen when clicking the file.  This is particularly
+                // important in the desktop app, where hovering a URL does
+                // not display the URL like it does in the web app.
+                url = $t(
```
The `url` variable first stored the actual `href` value of the link, and it is now storing the "tooltip label" — which is not a good practice and the variable name doesn't correspond to it's content. Lets use another variable to store the tooltip label, leaving the `url` untouched.

reharsh replied:
using `tooltip_label` now instead...

sayamsamal commented on web/src/message_list_tooltips.ts:
```diff
@@ -349,6 +352,60 @@ export function initialize(): void {
+            } else {
+                // We show the tooltip in canonical form for the
+                // destination internal link is pointing to.
+                const syntax = (): string => {
```
I don't see the particular need for a nested function here, the inner content could just as well be flattened outside.

reharsh replied:
done!

sayamsamal commented on web/src/message_list_tooltips.ts:
```diff
@@ -349,6 +352,60 @@ export function initialize(): void {
+                const format = syntax();
+                url = `Go to ${format}`;
+            }
+            instance.setContent(parse_html(render_external_inline_url_tooltip({url})));
```
We are using `external_inline_url_tooltip` Handlebars template for internal links as well, so the naming of the file is wrong? `message_inline_url_tooltip` could be a better naming for the template.

reharsh replied:
done!

sayamsamal commented on web/templates/external_inline_url_tooltip.hbs:
```diff
@@ -0,0 +1,3 @@
+<div id="external_inline_url_tooltip">
+    <div class="tooltip-inner-content italic">{{url}}</div>
+</div>
```
Although not an immediate issue, since we probably won't be having multiple instances of this tooltip at the same time in the DOM; it would be a standard practice to use a class name here so as to not have a invalid HTML markup incase of multiple elements in the DOM with the same `ID`.

sayamsamal replied:
Similar to the filename, we should also rename the `external_inline_url_tooltip` string used here.

reharsh replied:
done!

sayamsamal commented on web/src/message_list_tooltips.ts:
```diff
@@ -349,6 +352,64 @@ export function initialize(): void {
         },
     });
 
+    message_list_tooltip(".message_content > p > a[data-message-link-type='external_inline_url']", {
```
I think we can remove the in between `p` tag, since an `a` tag inside the message content should be specific enough.

reharsh replied:
done!

sayamsamal commented on web/src/message_list_tooltips.ts:
```diff
@@ -349,6 +352,64 @@ export function initialize(): void {
+                    const topic = stream_topic_name.topic_name;
+                    format = topic ? `#${channel} > ${topic}` : `#${channel}`;
+                } else {
+                    format = "this";
```
"Go to this" doesn't really convey anything and is rather confusing. If we are unable to determine if the internal link is either a user uploaded file, stream or stream+topic at this stage, we should probably just show the full url instead.

sayamsamal commented:
Was the decision anywhere to show the tooltip as italic? If I remember correctly, we only show the supporting text inside the tooltip as italic. For eg,
![Screenshot 2025-04-26 at 1 40 38 AM](https://github.com/user-attachments/assets/b175f99c-8e94-4c01-9ccf-917a934e1ee1)

sayamsamal commented:
Reading https://github.com/zulip/zulip/pull/33039#issuecomment-2590512516 and https://github.com/zulip/zulip/pull/33039#issuecomment-2605598608, I think @alya mentioned having no delay on the internal links and having a longer delay on the external links. The current implementation seems to be doing the opposite.

reharsh commented:
Actually I think showing the tooltips for external links (without any delay) is our goal so, that as alya mentioned in issue:

> This will help users notice when the URL for the link is not one they expect.

and we show tooltips for internal links (we know are safe) with longer delay?

sayamsamal commented:
@alya could you comment on this?

sayamsamal commented:
@reharsh left you a simple feedback, LGTM otherwise and we can go ahead with the integration review.

alya commented:
The canonical format for channel links is `#design`, not `# design`.

alya commented:
I think my intention was no delay on any links that lead to somewhere other than what it looks like, with no behavior difference between internal and external links.

alya commented:
@reharsh You can tag the PR for integration review once these points have been addressed.

reharsh commented:
@zulipbot remove "maintainer review"
@zulipbot add "integration review"

timabbott commented on web/src/postprocess_content.ts:
```diff
@@ -166,35 +167,26 @@ export function postprocess_content(html: string): string {
                     inline_image.classList.add("landscape-thumbnail");
                 }
             }
-        } else {
```
Why did you pull this out of the `else` clause? I think this may result in potentially both blocks running for images; is that intended?

timabbott commented on web/src/message_list_tooltips.ts:
```diff
@@ -349,6 +352,65 @@ export function initialize(): void {
+            instance.destroy();
+        },
+        placement: "bottom",
+    });
```
I'm not sure I see why you have this separate function with `external_inline_url` as a type -- why can't you just do all the calculations inside the `onShow` function?

timabbott commented on web/src/postprocess_content.ts:
```diff
@@ -166,35 +167,26 @@ export function postprocess_content(html: string): string {
+            !STANDARD_INTERNAL_URLS.has(elt.getAttribute("class") ?? "") &&
+            elt.textContent?.trim() !== elt.href.replace(/\/$/, "")
+        ) {
+            elt.setAttribute("data-message-link-type", "internal_named_link");
```
This conditions are incredibly hard to understand. I suggest defining some variables for values like `elt.textContent?.trim() !== elt.href.replace(/\/$/, "")` and `STANDARD_INTERNAL_URLS.has(elt.getAttribute("class") ?? ""` to make this code readable.
